django/core/paginator.py

```diff
@@ -107,10 +107,11 @@ def page_range(self):
 
     def _check_object_list_is_ordered(self):
         """
-        Warn if self.object_list is unordered (typically a QuerySet).
+        Warn if self.object_list is unordered (typically a QuerySet)
+        and non-empty.
         """
         ordered = getattr(self.object_list, 'ordered', None)
-        if ordered is not None and not ordered:
+        if ordered is not None and not ordered and self.object_list:
```

Comment:
Using `bool` on a queryset is going to be evaluating it in memory which defeats the purpose of pagination.

Reply:
Is there some way to check whether the object_list is empty without queries?
Could you give me some advice?

Reply:
@weijunji I think we'll need to introduce a `QuerySet.empty` property that does `return self.query.is_empty()` to prevent coupling the paginator logic to the ORM through `isinstance(query, EmptyQuerySet)` checks. Once this is done the following should work.

```python
ordered = getattr(self.object_list, 'ordered', None)
try:
    empty = self.object_list.empty
except AttributeError:
    empty = not self.count()
if ordered is not None and not ordered and not empty:
    ...
```

Reply:
@charettes  But it also can't pass this test:
```python
    def test_paginator_init_queries(self):
        """
        There should be none quesies while init the Paginator.
        """
        obj = Article.objects.all().order_by('id')
        self.assertNumQueries(0, Paginator, obj, 5)
```
Can we just deal with the Article.objects.none(), not solve the QuerySet which result is empty? Such as:
```python
from django.db.models.query import EmptyQuerySet
...
if ordered is not None and not ordered and not isinstance(self.object_list, EmptyQuerySet):
...
```

Reply:
@charettes And it seems that I couldn't find this property named `empty` in https://docs.djangoproject.com/en/dev/ref/models/querysets/#none
Is there some mistake?

Reply:
As mentioned in my previous comment we want to avoid coupling the paginator with the ORM, that translates into avoid imports from `django.db` in `django.core.paginator. That's the reason why we use duck typing to determine whether or not `object_list` is ordered instead of doing `ordered = object_list.ordered if isinstance(object_list, QuerySet) else None`.

> But it also can't pass this test...

Also mentioned above, the code won't work unless you introduce a `queryset.empty` property that proxies `self.query.is_empty()`. I suggested doing that to avoid embedding abstraction details about the ORM into this method and thus cause indirect coupling.

e.g. instead of doing the following

```python
try:
    empty = self.object_list.query.is_empty()
except AttributeError:
    empty = None
```

Reply:
@charettes Ok, I understand. I will try to add this property. Thanks a lot.

Reply:
@charettes why there is a check not successful? 
```
...
Checking out Revision 36c19e7700877795816ef09e87f92f24ba6d1dad (refs/remotes/origin/pull/9479/merge)
 > C:\Program Files\Git\bin\git.exe config core.sparsecheckout # timeout=10
 > C:\Program Files\Git\bin\git.exe checkout -f 36c19e7700877795816ef09e87f92f24ba6d1dad
FATAL: Could not checkout 36c19e7700877795816ef09e87f92f24ba6d1dad
hudson.plugins.git.GitException: Command "C:\Program Files\Git\bin\git.exe checkout -f 36c19e7700877795816ef09e87f92f24ba6d1dad" returned status code 128:
stdout: 
stderr: fatal: reference is not a tree: 36c19e7700877795816ef09e87f92f24ba6d1dad
...
```
This problem is as same as https://github.com/kubernetes/kubernetes/issues/27462
I think there are something wrong with the github PR. How can I fix this problem?